plugins/radare2/radare2_main.ml

```diff
@@ -0,0 +1,90 @@
+open Bap_core_theory
+open Core_kernel
+open Bap_future.Std
+open Bap.Std
+include Self()
+
+open KB.Syntax
+
+let agent =
+  KB.Agent.register ~package:"bap.std" "radare2-symbolizer"
+    ~desc:"extracts symbols radare2"
+
+let provide_roots funcs =
+  let promise_property slot =
+    KB.promise slot @@ fun label ->
+    KB.collect Theory.Label.addr label >>| function
+    | None -> None
+    | Some addr ->
+      let addr = Bitvec.to_bigint addr in
+      Option.some_if (Hashtbl.mem funcs addr) true in
+  promise_property Theory.Label.is_valid;
+  promise_property Theory.Label.is_subroutine
+
+exception Radare2_failed
+
+let provide_radare2 file = 
+  let funcs = Hashtbl.create (module struct
+      type t = Z.t
+      let compare = Z.compare and hash = Z.hash
+      let sexp_of_t x = Sexp.Atom (Z.to_string x)
+    end) in
+    let accept name addr = Hashtbl.set funcs addr name in
+    let extract name json = let open Yojson in
+      match json with
+      | `Assoc list -> 
+        (match List.find list ~f:(fun (key, _) -> String.equal key name) with
+        | Some (_, v) -> Some v
+        | _ -> None)
+      | _ -> None in
+    try
+     let symbol_list = let open Yojson in
+      match R2.with_command_j "isj" file with
+      | `List list -> Some list
+      | _ -> None in
+     let strip str = let open String in
+      match chop_prefix str ~prefix:"sym.imp." with
+      | Some str -> str
+      | None -> str in
+     let open Yojson in
+     let to_string json = 
+      match json with
+      | `String str -> Some str
+      | _ -> None in
+     let to_int json = 
+      match json with
+      | `Int i -> Z.of_int i |> Some
+      | `Intlit s -> Z.of_string s |> Some
+      | _ -> None in
+     Option.(symbol_list >>| List.fold ~init:() ~f:(fun () symbol -> 
+            both
+            (extract "name" symbol >>= to_string >>| strip) 
+            (extract "vaddr" symbol >>= to_int) 
+            |> value_map ~default:() ~f:(fun (name, addr) -> accept name addr)
+     ) |> value ~default:());
```

Comment:
This expression is overly complicated for the amount of work it is doing. A few suggestions on improving it.

1. `List.fold ~init:() ~f:(fun () x -> f x)` is better expressed as `List.iter ~f`
2. `Option.value (f x) ~default:()` is better expressed as `Option.iter ~f`
3. `Option.value_map ~default:() ~f` is better expressed as `Option.iter ~f`

I also suggest that instead of creating `to_string` and `to_int` parsing functions, which are only used once and do not really provide much abstraction, create two more specialized `extract_name` and `extract_addr` helpers, so that at the end the code could be written as,
```ocaml
List.iter symbol_list ~f:(fun s -> match extract_name s, extract_addr s with
  | Some name, Some addr -> accept name addr
  | _ -> debug "skipping %a" Yojson.pp s)
```


```